Better reaction to half-initialized GBPTree index file

tinwelint opened the pull request:
If creation of a GBPTree index file is interrupted mid-way, e.g. by process
crashing or something similar then the index file will exist, but its
meta pages empty or half-initialized. Previously only a missing index file
was interpreted correctly by throwing IOException, such that the layer
immediately above GBPTree would recognized that and rebuild that index or similar,
whereas the other cases of half-initialized meta pages would throw other
types of internal exceptions and not be properly handled by above layer.

This commit improves on this situation to interpret one or more missing or
all-zero meta pages the same way, i.e. by throwing proper IOException.

tinwelint committed:
Better reaction to half-initialized GBPTree index file

If creation of a GBPTree index file is interrupted mid-way, e.g. by process
crashing or something similar then the index file will exist, but its
meta pages empty or half-initialized. Previously only a missing index file
was interpreted correctly by throwing IOException, such that the layer
immediately above GBPTree would recognized that and rebuild that index or similar,
whereas the other cases of half-initialized meta pages would throw other
types of internal exceptions and not be properly handled by above layer.

This commit improves on this situation to interpret one or more missing or
all-zero meta pages the same way, i.e. by throwing proper IOException.

burqen commented on community/index/src/test/java/org/neo4j/index/internal/gbptree/GBPTreeTest.java:
```diff
@@ -684,6 +686,75 @@ public void readHeaderMustThrowIfFileDoesNotExist() throws Exception
     }
 
+    @Test
+    public void readHeaderMustThrowIOExceptionIfFileIsEmpty() throws Exception
```
Really sweet testing!
Can we also verify that constructor behave the same way as `readHeader` here?

tinwelint replied:
Oh of course, good call. I'll duplicate these tests for the constructor!

burqen reviewed (changes requested):
This is sweet 👍  I think we should unite

burqen reviewed (changes requested):
This is sweet! 👍 
I think we should also align constructor to behave in the same way and add testing for that. Constructor does not use `readHeader`.

Would also be neat if we could verify intended behaviour in some test. The population restart that is.

tinwelint commented:
Right about the testing... the test seeing this also saw all these three issues, so they spawned from that and that test could see them, but of course if that test changes then that accidental testing goes away. I can add specific ITs for this.

burqen commented:
@tinwelint That would be awesome!

tinwelint committed:
More testing for opening a partially created GBPTree

This uncovered paths that weren't guarded when opening a GBPTree normally
via its constructor (previous testing was only for its static readHeader method).
Also added GBPTreePartialCreateFuzzIT that, through randomized testing,
has the ability to uncover even more cases in the future.

tinwelint commented:
@burqen updated
